content/gotraining.slide

```diff
@@ -212,21 +224,279 @@ Execute with
+	2db0f12 fixed two build-breaking issues: + reverted ClassMetadataReadingVisitor to revision 794 + eliminated ConfigurationPostProcessorTests until further investigation determines why it causes downstream tests to fail (such as the seemingly unrelated ClassPathXmlApplicationContextTests)
+	147709f Tweaks to package-info.java files
+	22b25e0 Consolidated Util and MutableAnnotationUtils classes into existing AsmUtils
+	7f96f57 polishing
+
+versus
+
+	$ git log --oneline -5 --author pwebb --before "Sat Aug 30 2014"
+
+	5ba3db6 Fix failing CompositePropertySourceTests
+	84564a0 Rework @PropertySource early parsing logic
+	e142fd1 Add tests for ImportSelector meta-data
+	887815f Update docbook dependency and generate epub
+	ac8326d Polish mockito usage
+
+
+* Recipe for great commit messages
+
+[[https://chris.beams.io/posts/git-commit/]]
+
+- Separate subject from body with a blank line
+- Limit the subject line to 60 characters
+- *Capitalize* the subject line
+- Do not end the subject line with a period
+- Use the *imperative* *mood* in the subject line
+- Wrap the body at 80 characters
+- Use the body to explain what and why vs. how
+
+
+* Two `git` commands for precision work
+
+- Use `git` `rebase` `-i` `COMMIT_HASH` to rework your commits if necessary
+- Use `git` `add` `-p` for two separate concerns addressed in the one file
+
+Don't be afraid!
+Try it out on a new, throwaway branch.
+
+
+* Pull Request (PR) recommendations (1/2)
+
+- Choose a meaningful *PR* *title*, it will become your Squash Merge message
+- Fill in *PR* *description* and reference an Issue for instance with #6
+- Review the *Files* *changed* tab - many reviewers start here
+- Review the *Commits* tab - some reviewers start here
+- Review the *Checks* tab - all checks must pass
+
+
+* Pull Request (PR) recommendations (2/2)
+
+- "Think of the reviewer: your code needs to be reviewable and that should be a prime concern when writing your code, commits and PRs." (@camh-anz)
+- Tag your PR Title with `[WIP]` if adjustments are needed
+- Rework fixup commits with `git` `rebase` `-i`  `COMMIT_HASH` and force push
+- Add reviewers and remove `[WIP]` tag when all of the above are met
+- Delete branch once merged, especially when using a single remote (no forks)
+
+
 * Go language fundamentals
 
 
+* Tutorials, references and tools
+
+- [[https://tour.golang.org/][Tour of Go]]
+- [[https://gobyexample.com/][Go by example]]
+- [[https://play.golang.org/][Go playground]]
+- [[https://golangbot.com/learn-golang-series/][Golangbot]]
+- [[https://stackoverflow.com/questions/tagged/go?sort=votes&pageSize=100][Stackoverflow]]
+
```

Comment:
Perhaps add https://golang.org/ref/spec - the definitive reference. I know I'm weird, but that's how I learn languages.

Reply:
Done.